src/main/kotlin/org/jitsi/nlj/rtp/codec/vp9/Vp9Parser.kt

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.nlj.rtp.codec.vp9
+
+import org.jitsi.nlj.MediaSourceDesc
+import org.jitsi.nlj.PacketInfo
+import org.jitsi.nlj.rtp.codec.VideoCodecParser
+import org.jitsi.nlj.util.StateChangeLogger
+import org.jitsi.rtp.extensions.toHex
+import org.jitsi.utils.logging2.Logger
+import org.jitsi.utils.logging2.createChildLogger
+
+/**
+ * Some [Vp9Packet] fields are not able to be determined by looking at a single VP9 packet (for example the scalability
+ * structure is only carried in keyframes).  This class updates the layer descriptions with information
+ * from frames, and also diagnoses packet format variants that the Jitsi videobridge won't be able to route.
+ */
+class Vp9Parser(
+    sources: Array<MediaSourceDesc>,
+    parentLogger: Logger
+) : VideoCodecParser(sources) {
+    private val logger = createChildLogger(parentLogger)
+
+    private val pictureIdState = StateChangeLogger("missing picture id", logger)
+    private val extendedPictureIdState = StateChangeLogger("missing extended picture ID", logger)
+    private var numSpatialLayers = -1
+
+    /** Encodings we've actually seen.  Used to clear out inferred-from-signaling encoding information. */
+    private val ssrcsSeen = HashSet<Long>()
+
+    override fun parse(packetInfo: PacketInfo) {
+        val vp9Packet = packetInfo.packetAs<Vp9Packet>()
+
+        ssrcsSeen.add(vp9Packet.ssrc)
+
+        if (vp9Packet.hasScalabilityStructure) {
+            // TODO: handle case where new SS is from a packet older than the
```

Comment:
need to address this?  it does seem possible

Reply:
I'm not sure yet how likely it'll be.  I'd need to keep a packet index tracker to validate it, so I want to know if I need the extra state.

Reply:
Ok, so is the plan to track it after merging?  If so, should we at least have a stat to track how often it's happening?

Reply:
If I had a stat I could just handle it - the tricky part is tracking the sequence number as an extended sequence number.

Thinking about it, would this be something that'd be sensible to add to `PacketInfo` maybe?

Reply:
Hm, I don't think I followed the requirement for the extended sequence number

Reply:
I need to figure out if the packet with the SS we've just received is older or newer than the previous packet with an SS -- but we don't want it to be confused by sequence number cycles, which means tracking extended sequence numbers.